client/state/data-layer/wpcom/read/tags/utils.js

```diff
@@ -0,0 +1,40 @@
+/**
+ * External dependencies
+ */
+import { map, compact, concat, isObject, isArray } from 'lodash';
+import { decodeEntities } from 'lib/formatting';
+
+/**
+ * Normalize response from the api so whether we get back a single tag or a list of tags
+ * we always pass forward a list
+ * Also transform the api response to be something more calypso-friendly
+ *
+ * @param  {Tag|Tags} apiResponse api response from the tags endpoint
+ * @return {Tags} An object containing all of the normalized tags in the format:
+ *  [
+ *    { id, displayName, url, title, slug },
+ *    ...
+ *  ]
+ */
+export function fromApi( apiResponse ) {
+	if ( ! apiResponse || ( ! apiResponse.tag && ! apiResponse.tags ) ) {
+		if ( process.env.NODE_ENV === 'development' ) {
+			throw new Error( 'bad api response for /read/tags' );
+		}
+		return [];
+	}
+
+	const tags = compact( concat(
+		[],
+		isObject( apiResponse.tag ) && apiResponse.tag,
+		isArray( apiResponse.tags ) && apiResponse.tags,
+	) );
+
+	return map( tags, tag => ( {
+		id: tag.ID,
+		displayName: decodeEntities( tag.display_name ),
+		url: `/tag/${ tag.slug }`,
+		title: decodeEntities( tag.title ),
+		slug: tag.slug.toLowerCase(),
+	} ) );
+}
```

Comment:
from: https://github.com/Automattic/wp-calypso/pull/11435#discussion_r104765854

this is a very imperative way of assigning `tags` and I can't say that I can follow it very easily

it appears like the API might return _either_ a single tag _or_ a list of tags having a different structure. is this right?

also a big point of the `fromApi` function is showing how the API format of data maps into Calypso data, by simply passing along the data here we hide that and don't benefit from it

```js
const validate = ( { tag, tags } ) => {
	if ( ! tag && ! Array.isArray( tags ) ) {
		throw new TypeError(
			`Expected to receive tag information from API but none was found.\n` +
			`Expected either 'tag' as a tag object (but found '${ typeof tag }' instead)\n` +
			`or 'tags' as a list of tag objects (but found '${ typeof tags }' instead)`
		);
	}
}

const fromApi = ( { tag, tags } ) => compact(
	[ tag, ...tags ].map( tag => ( {
		id: tag.ID,
		slug: tag.slug,
		title: tag.title,
		displayName: tag.display_name,
		isFollowing: true,
	} ) )
);

import { tap } from 'lodash';

tags: fromApi( tap( data, validate ) )
```